muxers/yamux/src/lib.rs

```diff
@@ -141,17 +160,22 @@ where
     ) -> Poll<Result<StreamMuxerEvent, Self::Error>> {
         let this = self.get_mut();
 
-        loop {
-            let inbound_stream = ready!(this.poll_inner(cx))?;
-
-            if this.inbound_stream_buffer.len() >= MAX_BUFFERED_INBOUND_STREAMS {
-                log::warn!("dropping {inbound_stream} because buffer is full");
-                drop(inbound_stream);
-                continue;
-            }
+        let inbound_stream = ready!(this.poll_inner(cx))?;
 
+        if this.inbound_stream_buffer.len() >= this.max_buffered_inbound_streams {
+            log::warn!("dropping {inbound_stream} because buffer is full");
+            drop(inbound_stream);
+        } else {
             this.inbound_stream_buffer.push_back(inbound_stream);
+
+            if let Some(waker) = this.inbound_stream_waker.take() {
+                waker.wake()
```

Comment:
The discussion above makes me think: is this muxer polled from multiple tasks? If it is, then `poll_inner` will probably switch out wakers all the time, making it non-deterministic which caller will eventually be woken. If not, then this extra wakeup makes little sense to me.

Reply:
At least in `rust-libp2p` production code, it is only polled from a single task but it is a public interface so there may be other consumers, https://github.com/libp2p/rust-libp2p/pull/2952 for example.